pandas/core/arrays/datetimelike.py

```diff
@@ -196,6 +196,71 @@ def astype(self, dtype, copy=True):
             return self._box_values(self.asi8)
         return super(DatetimeLikeArrayMixin, self).astype(dtype, copy)
 
+    # ------------------------------------------------------------------
+    # ExtensionArray Interface
+    # TODO:
+    #   * _from_sequence
+    #   * argsort / _values_for_argsort
+    #   * _reduce
+
+    def unique(self):
+        result = unique1d(self.asi8)
+        return type(self)(result, dtype=self.dtype)
+
+    def _validate_fill_value(self, fill_value):
+        """
+        If a fill_value is passed to `take` convert it to an i8 representation,
+        raising ValueError if this is not possible.
+
+        Parameters
+        ----------
+        fill_value : object
+
+        Returns
+        -------
+        fill_value : np.int64
+
+        Raises
+        ------
+        ValueError
+        """
+        raise AbstractMethodError(self)
+
+    def take(self, indices, allow_fill=False, fill_value=None):
+        if allow_fill:
+            fill_value = self._validate_fill_value(fill_value)
+
+        new_values = take(self.asi8,
+                          indices,
+                          allow_fill=allow_fill,
+                          fill_value=fill_value)
+
+        return type(self)(new_values, dtype=self.dtype)
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        # for TimedeltaArray and PeriodArray; DatetimeArray overrides
```

Comment:
The problem being that DatetimeArray needs to pass through `tz` info?

For PeriodArray at least (haven't checked TimedeltaArray) you should able to implement `_concat_same_type` just in terms of `.dtype`. It's hashable and can be passed to `PeriodArray.__init__`.

Reply:
Good call

Reply:
As opposed to PeriodArray, `freq` is not part of the dtype for DatetimeArray/TimedeltaArray, so I am not sure this check for freq should be done here.

Reply:
IIRC, the PeriodArray constructor allows duplicate `freq` and `dtype`, as long as the match, so

```python
PeriodArray(data, freq='H', dtype=PeriodDtype("H")
```

should be fine.

Reply:
Yah, we added dtype to the PeriodArray constructor specifically so that type(self)(values, freq=self.freq, dtype=self.dtype) would be valid for all three TDA/DTA/PA classes

Reply:
Sorry, that's not what I was meaning. The difference is the meaning of `freq` for PeriodArray vs DatetimeArray. For PeriodArray it is part of the dtype and it is defining how the stored integers are interpreted (and thus need to match to just concatenate those integers), but for DatetimeArray it is simply an informative attribute telling you about the regularity of the Array, but not essential to describe it. So I would assume that `_concat_same_type` needs to handle different arrays with different freqs for DatetimeArray.